phantazm-neuron-minestom/src/main/java/com/github/phantazmnetwork/neuron/bindings/minestom/entity/GroundController.java

```diff
@@ -0,0 +1,170 @@
+ */
+public class GroundController implements Controller {
+    private final LivingEntity entity;
+    private final double step;
+
+    private boolean jumping;
+
+    private double lastH = -1;
+    private double lastB = -1;
+    private double lastG = -1;
+
+    private double lastJumpVelocity = -1;
+
+    /**
+     * Creates a new GroundController managing the provided entity, using the given step distance and walk speed.
+     * @param entity the entity managed by this controller
+     * @param step the maximum distance the entity may "step up" blocks
+     */
+    public GroundController(@NotNull LivingEntity entity, float step) {
+        this.entity = Objects.requireNonNull(entity, "entity");
+        this.step = step;
+    }
+
+    @Override
+    public double getX() {
+        return entity.getPosition().x();
+    }
+
+    @Override
+    public double getY() {
+        return entity.getPosition().y();
+    }
+
+    @Override
+    public double getZ() {
+        return entity.getPosition().z();
+    }
+
+    //this method's code is adapted from net.minestom.server.entity.pathfinding.Navigator#moveTowards(Point, double)
+    @Override
+    public void advance(@NotNull Node current, @NotNull Node target) {
+        Vec3I targetPos = target.getPosition();
+        Pos entityPos = entity.getPosition();
+
+        double exactTargetY = targetPos.getY() + target.getYOffset();
+
+        double dX = (targetPos.getX() + target.getXOffset()) - entityPos.x();
+        double dY = exactTargetY - entityPos.y();
+        double dZ = (targetPos.getZ() + target.getZOffset()) - entityPos.z();
+
+        //slows down entities when they reach their position
+        double distSquared = dX * dX + dY * dY + dZ * dZ;
+        double speed = entity.getAttributeValue(Attribute.MOVEMENT_SPEED);
+        if (speed > distSquared) {
+            speed = distSquared;
+        }
+
+        double radians = Math.atan2(dZ, dX);
+        double vX = Math.cos(radians) * speed;
+        double vZ = Math.sin(radians) * speed;
+
+        //make sure speedX and speedZ cannot extend past the target
+        double speedX = Math.copySign(Math.min(Math.abs(vX), Math.abs(dX)), dX);
+        double speedZ = Math.copySign(Math.min(Math.abs(vZ), Math.abs(dZ)), dZ);
+
+        if(jumping) {
+            if(entityPos.y() > exactTargetY) {
+                //jump completed successfully
+                entity.refreshPosition(CollisionUtils.handlePhysics(entity, new Vec(speedX, 0, speedZ)).newPosition()
+                        .withView(PositionUtils.getLookYaw(dX, dZ), 0));
+                entity.setVelocity(Vec.ZERO);
+                jumping = false;
+                return;
+            }
+            else if(entity.isOnGround()) {
+                //jump failed (we're back on the ground)
+                jumping = false;
+            }
+            else {
+                //still jumping
+                return;
```

Comment:
the dreaded else return

Reply:
i believe it is actually necessary here, as we still want to run the rest of the function if the execution falls through the else if